lua/lspconfig/util.lua

```diff
@@ -21,6 +21,21 @@ M.default_config = {
 -- global on_setup hook
 M.on_setup = nil
 
+---@param filetype string the filetype to check (can be a compound, dot-separated filetype; see |'filetype'|)
+---@param expected string|string[] the filetype(s) to match against
+---@return boolean
+function M.ft_matches(filetype, expected)
```

Comment:
there has vim.fs module.

Reply:
sorry, can you explain? I don't see anything in `vim.fs` that does this. vim.fs seems to deal mostly with filesystem operations. I'm not actually reading files here, I am only splitting and comparing `filetype` strings

Reply:
okay see vim.filetype moudle. vim.filetype.match will return the filetype if it already in default detect table or register by vim.filetype.add . we don't want add more utils function now

Reply:
`vim.filetype.match` takes an existing buffer and returns the `filetype` (using the rules defined in `filetype.lua`). the result of that might be (for example) `c.doxygen`. see `|'filetype'|`:

```vimhelp
						*'filetype'* *'ft'*
'filetype' 'ft'		string	(default "")
			local to buffer  |local-noglobal|
...
	Setting this option to a different value is most useful in a modeline,
	for a file for which the file type is not automatically recognized.
	Example, for in an IDL file: >c
		/* vim: set filetype=idl : */
<	|FileType| |filetypes|
	When a dot appears in the value then this separates two filetype
	names.  Example: >c
		/* vim: set filetype=c.doxygen : */
<	This will use the "c" filetype first, then the "doxygen" filetype.
	This works both for filetype plugins and for syntax files.  More than
	one dot may appear.
	This option is not copied to another buffer, independent of the 's' or
	'S' flag in 'cpoptions'.
	Only normal file name characters can be used, `/\*?[|<>` are illegal.
```

In other words:

```lua
local ft = vim.filetype.match({buf=0})
ft == "c.doxygen"
util.ft_matches(ft, 'idl') == false
util.ft_matches(ft, 'c') == true
util.ft_matches(ft, 'doxygen') == true
```

Reply:
you can also pass filename  into match

Reply:
yes, `vim.filetype.match` can take a filename and give us a filetype. but what we want is to take *that* filetype and check it against a list of filetypes (e.g. the ones supported by a particular lsp server). 

i.e. 

```lua
local ft = vim.filetype.match({ filename = 'main.lua' })
ft == "lua"
local filetypes_supported_by_current_client = { "lua", "sh", "c" }
util.ft_matches(ft, filetypes_supported_by_current_client) == true
```

Usually this would just be a `==` comparison but because a filetype can contain multiple filetypes (e.g. `c.doxygen`), we have to split it first.

```lua
local ft = "yaml.ansible" -- this is not a filename! it is a _compound_ filetype
util.ft_matches(ft, "ansible") == true
util.ft_matches(ft, "yaml") == true
```

notice that `vim.filetype.match` doesn't help us do this comparison! it takes a buffer/file/filepath and gives us a filetype. but we don't need that. we **have** a filetype, it is provided by the caller. We are trying to match that against a list of acceptable filetypes, also provided by the caller.

Reply:
does register a filetype  by using vim.filetype.add for a special file not make sense in user side

Reply:
A completely unique type is not desirable because we want to use the existing behaviors defined for the primary filetype. see [mfussenegger/nvim-ansible](https://github.com/mfussenegger/nvim-ansible/blob/cb5d0cc79c1864b72d5ae40a926a1176b18f1b65/ftdetect/ansible.lua#L23) or [pearofducks/ansible-vim](https://github.com/pearofducks/ansible-vim/blob/master/ftdetect/ansible.vim#L36). They both `set ft=yaml.ansible`.

You *want* ansible scripts to have the yaml filetype because that's what they are; we want (neo)vim to keep recognizing them as yaml files. if we set `ft=ansible` only, we lose the syntax highlighting/`match_words`/etc defined for yaml files. 

But the plugin additionally [sets various useful options](https://github.com/mfussenegger/nvim-ansible/tree/main?tab=readme-ov-file#usage) like `keywordprg`/`path`/`iskeyword`/`isfname` for `ansible` file types. Hence the reason for multiple file types.

The problem is that right now, nvim-lspconfig diverges from standard vim behavior because filetype matching operates on the entire `vim.bo.filetype` as a whole, rather than splitting it into individual file types. And every server has to be configured for every possible combination of multiple filetypes. see:

https://github.com/neovim/nvim-lspconfig/blob/a3d9395455f2b2e3b50a0b0f37b8b4c23683f44a/lua/lspconfig/server_configurations/yamlls.lua#L4-L6

(notice that `yaml.ansible` is not included.

With this change, you can instead just specify `yaml`, and we'll match as long as *one* of your filetypes matches `yaml`